crates/evm/coverage/src/analysis.rs

```diff
@@ -511,127 +466,36 @@ impl SourceAnalyzer {
-                    }
-                }
-
-                contracts_libraries
-            } {
-                // get items of each contract or library
-                if let Some(items) = self.contract_items.get(contract_or_lib) {
-                    item_ids.extend(items.iter());
-                }
-            }
-
-            // If there are no items for this contract, then it was most likely filtered
-            if !item_ids.is_empty() {
-                flattened.insert(contract_id.clone(), item_ids);
-            }
-        }
-
-        Ok(SourceAnalysis { items: self.items.clone(), contract_items: flattened })
-    }
-
-    fn analyze_contracts(&mut self) -> eyre::Result<()> {
         for contract_id in self.contracts.keys() {
-            // Find this contract's coverage items if we haven't already
-            if !self.contract_items.contains_key(contract_id) {
-                let ContractVisitor { items, base_contract_node_ids, .. } = ContractVisitor::new(
-                    contract_id.source_id,
-                    self.sources.get(&contract_id.source_id).unwrap_or_else(|| {
-                        panic!(
-                            "We should have the source code for source ID {}",
-                            contract_id.source_id
-                        )
-                    }),
-                    contract_id.contract_name.clone(),
-                )
-                .visit(
-                    self.contracts
-                        .get(contract_id)
-                        .unwrap_or_else(|| {
-                            panic!("We should have the AST of contract: {contract_id:?}")
-                        })
-                        .clone(),
-                )?;
-
-                let is_test = items.iter().any(|item| {
-                    if let CoverageItemKind::Function { name } = &item.kind {
-                        name.is_test()
-                    } else {
-                        false
-                    }
-                });
-
-                // Record this contract's base contracts
-                // We don't do this for test contracts because we don't care about them
-                if !is_test {
-                    self.contract_bases.insert(
-                        contract_id.clone(),
-                        base_contract_node_ids
-                            .iter()
-                            .filter_map(|base_contract_node_id| {
-                                self.contract_ids.get(base_contract_node_id).cloned()
-                            })
-                            .collect(),
-                    );
-                }
-
-                // For tests and contracts with no items we still record an empty Vec so we don't
-                // end up here again
-                if items.is_empty() || is_test {
-                    self.contract_items.insert(contract_id.clone(), Vec::new());
+            let ContractVisitor { items, .. } = ContractVisitor::new(
```

Comment:
w/o the `if !self.contract_items.contains_key(contract_id) {` check, will this not lead to re-analyzing contracts? and possibly duping coverage items?

Reply:
We're walking only `ContractDefinition` nodes from `self.contracts`, so it should be fine as long as there is no duplicating `ContractId` keys

Reply:
i think thats a fair assumption